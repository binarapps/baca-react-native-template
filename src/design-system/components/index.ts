export * from './Absolute'
export * from './AlertBox'
export * from './Box'
export * from './BoxWithShadow'
export * from './Button'
export * from './Center'
export * from './Column'
export * from './GradientBox'
export * from './Row'
export * from './Spacer'
export * from './Slider'
export * from './Text'
export * from './Touchables'
export * from './ScrollView'
export { LinearGradient as Gradient } from 'expo-linear-gradient'

export * from './CheckboxButton'
export * from './FormErrorMessage'
export * from './FormLabel'
export * from './Icon'
export * from './Input'
export * from './RadioButton'
export * from './Loader'
export * from './Select'
export * from './types'
export * from './RenderComponentWithExample'
