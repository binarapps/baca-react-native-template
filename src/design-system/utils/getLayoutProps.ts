const layoutPropsKeys = [
  'm',
  'margin',
  'mt',
  'marginTop',
  'mr',
  'marginRight',
  'mb',
  'marginBottom',
  'ml',
  'marginLeft',
  'mx',
  'my',
  'p',
  'padding',
  'pt',
  'paddingTop',
  'pr',
  'paddingRight',
  'pb',
  'paddingBottom',
  'pl',
  'paddingLeft',
  'px',
  'py',
]

export const getLayoutProps = <T extends object>(props: T) => {
  const layoutProps = Object.fromEntries(
    Object.entries(props).filter(([key]) => layoutPropsKeys.includes(key))
  )

  const restProps = Object.fromEntries(
    Object.entries(props).filter(([key]) => !layoutPropsKeys.includes(key))
  )

  return { layoutProps, restProps }
}
