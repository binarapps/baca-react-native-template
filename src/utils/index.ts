export * from './alert'
export * from './checkForUpdates'
export * from './convertEmToNumber'
export * from './getApiError'
export * from './getNavigator'
export * from './getColorValue'
export * from './handleFormErrors'
export * from './hapticImpact'
export * from './hex2rgba'
export * from './noop'
export * from './removeFalsyProperties'
export * from './secureStore'
export * from './showToast'
export * from './wait'
