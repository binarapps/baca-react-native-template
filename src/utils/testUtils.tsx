import { PortalProvider } from '@gorhom/portal'
import { render, RenderAPI } from '@testing-library/react-native'
import { PropsWithChildren, ReactElement } from 'react'
import { I18nextProvider } from 'react-i18next'

import i18n from '@/i18n/i18nForTests'
import { ColorSchemeProvider } from '@/providers/ColorSchemeProvider'

type RenderOptions = Parameters<typeof render>[1]

const ProvidersWrapper: React.FC<PropsWithChildren> = ({ children }) => (
  <PortalProvider>
    <ColorSchemeProvider>
      <I18nextProvider i18n={i18n}>{children}</I18nextProvider>
    </ColorSchemeProvider>
  </PortalProvider>
)

const customRender = (ui: ReactElement, options?: RenderOptions): RenderAPI => {
  return render(ui, { wrapper: ProvidersWrapper, ...options })
}

export * from '@testing-library/react-native'
export { customRender as render }
