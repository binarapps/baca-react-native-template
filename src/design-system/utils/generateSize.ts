import { ViewStyle, DimensionValue, Animated } from 'react-native'

import { SizingValue } from '@/design-system'
import { size } from '@/design-system/config'

const isDimensionValue = (value?: DimensionValue) => {
  // Dimension accepts only `percentage` strings that can be converted to numbers
  if (typeof value === 'string' && value.endsWith('%') && !isNaN(parseFloat(value.split('%')[0]))) {
    return true
  }

  if (typeof value === 'number') {
    return true
  }

  if (value instanceof Animated.Value || value instanceof Animated.ValueXY) {
    return true
  }
}

/**
 * Generates a ViewStyle object with a given spacing or dimension value and key.
 * @param value The sizing (based on theme) or dimension value to generate the ViewStyle object with.
 * @param key The key of the ViewStyle object to assign the generated value to.
 * @returns A ViewStyle object with the generated value assigned to the given key, or undefined if the value or key is not provided or the generated value is invalid.
 * @example
 *
 * valid values are:
 * - '0'
 * - '1'
 * - '2'
 * - '1/2'
 * - 'full'
 * - '256px'
 * - '50%'
 * - 'auto'
 * - 'Animated.Value(1)'
 * - 'Animated.ValueXY({ x: 1, y: 1 })'
 * - 2137
 * - 1337
 *
 * invalid values are:
 * -  {}
 * - '2137em'
 * - '2137rem'
 * - '2137vw'
 * - '2137vh'
 * - 'some string'
 */
export const generateSize = (value?: SizingValue, key?: keyof ViewStyle): ViewStyle | undefined => {
  if (!value || !key) return

  // @ts-expect-error: native base color literal
  const spaceValue = size[value]

  // strict check because 0 is falsy
  if (spaceValue !== undefined) {
    return { [key]: spaceValue }
  }

  // strict check because 0 is falsy
  if (typeof value === 'number' && spaceValue === undefined) {
    return { [key]: value }
  }

  if (typeof value === 'string' && value.endsWith('px')) {
    return { [key]: parseInt(value.split('px')[0], 10) }
  }

  if (isDimensionValue(value as DimensionValue)) {
    return { [key]: value }
  }

  if (value === 'auto') {
    return { [key]: 'auto' }
  }

  return undefined
}
