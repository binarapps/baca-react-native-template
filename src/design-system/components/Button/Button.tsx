import {
  useMemo,
  memo,
  useCallback,
  forwardRef,
  ForwardRefExoticComponent,
  PropsWithoutRef,
  RefAttributes,
} from 'react'
import {
  Pressable,
  PressableProps,
  PressableStateCallbackType,
  StyleSheet,
  ViewStyle,
  TextStyle,
  StyleProp,
  View,
  Platform,
} from 'react-native'

import {
  ButtonSize,
  ButtonVariant,
  buttonSizeVariants,
  buttonVariants,
  getButtonShadowStyle,
} from '../../config'
import { generateStyledComponent } from '../../utils'
import { Icon } from '../Icon'
import { Loader } from '../Loader'
import { Row } from '../Row'
import { Text } from '../Text'
import { useHover } from '../Touchables/useHover'
import { StyledProps } from '../types'

import { useTheme } from '@/hooks'
import { IconNames } from '@/types'
import { getColorValue } from '@/utils'

export type ButtonProps = StyledProps &
  PressableProps & {
    disabled?: boolean
    leftElement?: JSX.Element
    leftIconName?: IconNames
    loaderElement?: JSX.Element
    loading?: boolean
    rightIconName?: IconNames
    size?: ButtonSize
    textStyle?: StyleProp<TextStyle>
    title?: string
    variant?: ButtonVariant
  }

const styles = StyleSheet.create({
  baseButton: {
    alignItems: 'center',
    borderRadius: 4,
    flexDirection: 'row',
    justifyContent: 'center',
  },
  baseText: {
    fontStyle: 'normal',
    fontWeight: '400',
  },
})

const RawButton = memo(
  forwardRef<View, ButtonProps>(
    (
      {
        children,
        disabled,
        leftElement,
        leftIconName,
        loading,
        rightIconName,
        size = 'md',
        style,
        textStyle,
        title,
        variant = 'Primary',
        ...props
      },
      ref
    ) => {
      const { colors } = useTheme()
      const { hoverProps, isHovered } = useHover()
      const { hoveredStyle, defaultStyle, disabledStyle } = useMemo(
        () => buttonVariants[variant],
        [variant]
      )

      const pressedStyles = useMemo<ViewStyle>(() => {
        return getButtonShadowStyle({ variant })
      }, [variant])

      const hoveredStyles = useMemo<ViewStyle>(
        () => ({
          backgroundColor: getColorValue({
            color: hoveredStyle.backgroundColor,
            colors,
          }),
          borderColor: getColorValue({
            color: hoveredStyle.borderColor!,
            colors,
          }),
          borderWidth: hoveredStyle.borderWidth,
        }),
        [colors, hoveredStyle.backgroundColor, hoveredStyle.borderColor, hoveredStyle.borderWidth]
      )

      const hoverColorStyle = useMemo<TextStyle>(
        () => ({
          color: getColorValue({
            color: hoveredStyle.color!,
            colors,
          }),
        }),
        [colors, hoveredStyle.color]
      )

      const defaultStyles = useMemo<ViewStyle>(
        () => ({
          backgroundColor: getColorValue({
            color: defaultStyle.backgroundColor,
            colors,
          }),
          borderColor: getColorValue({
            color: defaultStyle.borderColor!,
            colors,
          }),
          borderWidth: defaultStyle.borderWidth,
        }),
        [colors, defaultStyle.backgroundColor, defaultStyle.borderColor, defaultStyle.borderWidth]
      )

      const defaultColorStyle = useMemo<TextStyle>(
        () => ({
          color: getColorValue({
            color: defaultStyle.color!,
            colors,
          }),
        }),
        [colors, defaultStyle.color]
      )

      const disabledStyles = useMemo<ViewStyle>(
        () => ({
          backgroundColor: getColorValue({
            color: disabledStyle.backgroundColor,
            colors,
          }),
          borderColor: getColorValue({
            color: disabledStyle.borderColor!,
            colors,
          }),
          borderWidth: disabledStyle.borderWidth,
        }),
        [
          colors,
          disabledStyle.backgroundColor,
          disabledStyle.borderColor,
          disabledStyle.borderWidth,
        ]
      )

      const disabledColorStyle = useMemo<TextStyle>(
        () => ({
          color: getColorValue({
            color: disabledStyle.color!,
            colors,
          }),
        }),
        [colors, disabledStyle.color]
      )
      const buttonSizeVariant = buttonSizeVariants[size]

      const buttonSizeStyle = useMemo<ViewStyle>(
        () => ({
          paddingHorizontal: buttonSizeVariant.paddingHorizontal,
          paddingVertical: buttonSizeVariant.paddingVertical,
        }),
        [buttonSizeVariant.paddingHorizontal, buttonSizeVariant.paddingVertical]
      )

      const pressableStyleFunction = useCallback(
        ({ pressed }: PressableStateCallbackType): StyleProp<ViewStyle> =>
          StyleSheet.flatten<ViewStyle>([
            styles.baseButton,
            {
              ...Platform.select({
                default:
                  pressed || isHovered ? { ...hoveredStyles, ...pressedStyles } : defaultStyles,
                web: pressed
                  ? { ...defaultStyles, ...pressedStyles }
                  : isHovered
                    ? hoveredStyles
                    : defaultStyles,
              }),
            },
            disabled && disabledStyles,
            loading && disabledStyles,
            buttonSizeStyle,
            typeof style === 'function' ? style({ pressed }) : style,
          ]),
        [
          buttonSizeStyle,
          defaultStyles,
          disabled,
          disabledStyles,
          hoveredStyles,
          isHovered,
          loading,
          pressedStyles,
          style,
        ]
      )

      const getIconColor = useCallback(
        ({ pressed }: PressableStateCallbackType): ColorNames | undefined => {
          if (disabled) return disabledStyle.color
          if (pressed || isHovered) return hoveredStyle.color
          return defaultStyle.color
        },
        [defaultStyle.color, disabled, disabledStyle.color, hoveredStyle.color, isHovered]
      )

      const pressableTextStyleFunction = useCallback(
        ({ pressed }: PressableStateCallbackType) =>
          StyleSheet.flatten([
            styles.baseText,
            { lineHeight: buttonSizeVariant.lineHeight },
            pressed ? hoverColorStyle : defaultColorStyle,
            disabled && disabledColorStyle,
            textStyle,
          ]),
        [
          buttonSizeVariant.lineHeight,
          hoverColorStyle,
          defaultColorStyle,
          disabled,
          disabledColorStyle,
          textStyle,
        ]
      )

      const iconElement = useCallback(
        (props: PressableStateCallbackType, iconName?: IconNames) => {
          return iconName ? (
            <Icon name={iconName} size={buttonSizeVariant.iconSize} color={getIconColor(props)} />
          ) : null
        },
        [buttonSizeVariant.iconSize, getIconColor]
      )

      const childrenElement = useCallback(
        (props: PressableStateCallbackType) => {
          if (title) {
            return (
              <Text
                allowFontScaling={false}
                selectable={false}
                style={pressableTextStyleFunction(props)}
                textAlign="center"
                variant={buttonSizeVariant.textVariant}
              >
                {title}
              </Text>
            )
          }
          if (typeof children === 'string') {
            return (
              <Text
                allowFontScaling={false}
                selectable={false}
                style={pressableTextStyleFunction(props)}
                textAlign="center"
                variant={buttonSizeVariant.textVariant}
              >
                {children}
              </Text>
            )
          }
          return <>{children}</>
        },

        [buttonSizeVariant.textVariant, children, pressableTextStyleFunction, title]
      )

      return (
        <Pressable
          accessibilityRole="button"
          disabled={disabled || loading}
          role="button"
          style={pressableStyleFunction}
          testID="baseButton"
          {...{ ...hoverProps, ref, ...props }}
        >
          {loading ? (
            <Loader
              color={colors.button.primary.bg}
              size={buttonSizeVariant.iconSize}
              type="default"
            />
          ) : (
            (props: PressableStateCallbackType) => (
              <Row alignItems="center" gap={buttonSizeVariant.iconGap}>
                {leftElement && leftElement}
                {leftIconName && iconElement(props, leftIconName)}
                {childrenElement(props)}
                {rightIconName && iconElement(props, rightIconName)}
              </Row>
            )
          )}
        </Pressable>
      )
    }
  )
)

type ButtonComposition = ForwardRefExoticComponent<
  PropsWithoutRef<ButtonProps> & RefAttributes<View>
> & {
  [key in ButtonVariant]: ForwardRefExoticComponent<
    PropsWithoutRef<ButtonProps> & RefAttributes<View>
  >
}
const Button = generateStyledComponent<ButtonProps, typeof Pressable>(
  RawButton
) as ButtonComposition

Button.displayName = 'Button'

const generateButtonVariant = (variant: ButtonVariant) =>
  forwardRef<View, ButtonProps>((props, ref) => <Button {...{ ...props, ref, variant }} />)

Button.Primary = generateButtonVariant('Primary')
Button.Primary.displayName = 'Button.Primary'

Button.PrimaryDestructive = generateButtonVariant('PrimaryDestructive')
Button.PrimaryDestructive.displayName = 'Button.PrimaryDestructive'

Button.SecondaryColor = generateButtonVariant('SecondaryColor')
Button.SecondaryColor.displayName = 'Button.SecondaryColor'

Button.SecondaryGray = generateButtonVariant('SecondaryGray')
Button.SecondaryGray.displayName = 'Button.SecondaryGray'

Button.SecondaryDestructive = generateButtonVariant('SecondaryDestructive')
Button.SecondaryDestructive.displayName = 'Button.SecondaryDestructive'

Button.TertiaryColor = generateButtonVariant('TertiaryColor')
Button.TertiaryColor.displayName = 'Button.TertiaryColor'

Button.TertiaryGray = generateButtonVariant('TertiaryGray')
Button.TertiaryGray.displayName = 'Button.TertiaryGray'

Button.TertiaryDestructive = generateButtonVariant('TertiaryDestructive')
Button.TertiaryDestructive.displayName = 'Button.TertiaryDestructive'

Button.LinkColor = generateButtonVariant('LinkColor')
Button.LinkColor.displayName = 'Button.LinkColor'

Button.LinkGray = generateButtonVariant('LinkGray')
Button.LinkGray.displayName = 'Button.LinkGray'

Button.LinkDestructive = generateButtonVariant('LinkDestructive')
Button.LinkDestructive.displayName = 'Button.LinkDestructive'

export { Button }
