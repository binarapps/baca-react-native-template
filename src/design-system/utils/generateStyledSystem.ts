import { ImageStyle, StyleProp, StyleSheet, TextStyle, ViewStyle } from 'react-native'

import { generateSize } from './generateSize'
import {
  FlexProps,
  SpacingProps,
  StyledProps,
  LayoutsProps,
  BordersProps,
  SizingProps,
  EffectsProps,
  BackgroundProps,
} from '../components/types'

import { getColorValue, hex2rgba, removeFalsyProperties } from '@/utils'

const generateFlexStyle = ({
  flex,
  justifyContent,
  alignItems,
  alignContent,
  alignSelf,
  flexDirection,
  flexGrow,
  flexBasis,
  flexShrink,
  flexWrap,
}: FlexProps): StyleProp<ViewStyle> => {
  const style: StyleProp<ViewStyle> = [
    !!flex && { flex },
    !!justifyContent && { justifyContent },
    !!alignItems && { alignItems },
    !!alignContent && { alignContent },
    !!alignSelf && { alignSelf },
    !!flexDirection && { flexDirection },
    !!flexGrow && { flexGrow },
    !!flexBasis && { flexBasis },
    !!flexShrink && { flexShrink },
    !!flexWrap && { flexWrap },
  ]
  return style.filter(Boolean)
}

const generateSpacingStyle = ({
  m,
  mt,
  ml,
  mb,
  mr,
  mx,
  my,
  p,
  pr,
  pt,
  pl,
  pb,
  px,
  py,
  gap,
}: SpacingProps): StyleProp<ViewStyle> => {
  const style: StyleProp<ViewStyle> = [
    generateSize(p, 'padding'),
    generateSize(px, 'paddingHorizontal'),
    generateSize(py, 'paddingVertical'),
    generateSize(pb, 'paddingBottom'),
    generateSize(pt, 'paddingTop'),
    generateSize(pl, 'paddingLeft'),
    generateSize(pr, 'paddingRight'),
    generateSize(m, 'margin'),
    generateSize(mx, 'marginHorizontal'),
    generateSize(my, 'marginVertical'),
    generateSize(mb, 'marginBottom'),
    generateSize(mt, 'marginTop'),
    generateSize(ml, 'marginLeft'),
    generateSize(mr, 'marginRight'),
    generateSize(gap, 'gap'),
  ]

  return style.filter(Boolean)
}

const generateBgStyle = (
  { bg, backgroundColor, backgroundOpacity = 1, bgOpacity = 1 }: BackgroundProps,
  colors: Colors
): StyleProp<ViewStyle> => {
  if (!bg && !backgroundColor) return undefined

  const color = getColorValue({ color: (bg || backgroundColor) as ColorNames, colors })
  if (color && (typeof bgOpacity === 'number' || typeof backgroundOpacity === 'number')) {
    return { backgroundColor: hex2rgba(color, (bgOpacity || backgroundOpacity) as number) }
  }

  return { backgroundColor: bg }
}

const generateSizingStyle = ({
  w,
  width,
  h,
  height,
  minW,
  minWidth,
  minH,
  minHeight,
  maxW,
  maxWidth,
  maxH,
  maxHeight,
}: SizingProps): StyleProp<ViewStyle> => {
  const style: StyleProp<ViewStyle> = [
    generateSize(w, 'width'),
    generateSize(width, 'width'),
    generateSize(h, 'height'),
    generateSize(height, 'height'),
    generateSize(minW, 'minWidth'),
    generateSize(minWidth, 'minWidth'),
    generateSize(minH, 'minHeight'),
    generateSize(minHeight, 'minHeight'),
    generateSize(maxW, 'maxWidth'),
    generateSize(maxWidth, 'maxWidth'),
    generateSize(maxH, 'maxHeight'),
    generateSize(maxHeight, 'maxHeight'),
  ]
  return style
}

const generateEffectsStyle = ({ opacity }: EffectsProps): StyleProp<ViewStyle> => {
  return {
    opacity,
  }
}

const generateBordersStyle = (
  {
    borderColor,
    borderRadius,
    borderStyle,
    borderWidth,
    borderBottomEndRadius,
    borderBottomLeftRadius,
    borderBottomWidth,
    borderLeftWidth,
    borderRightWidth,
    borderTopEndRadius,
    borderTopLeftRadius,
    borderTopWidth,
  }: BordersProps,
  colors: Colors
): StyleProp<ViewStyle> => {
  return [
    !!borderColor && { borderColor: getColorValue({ color: borderColor, colors }) },
    {
      borderBottomEndRadius,
      borderBottomLeftRadius,
      borderBottomWidth,
      borderLeftWidth,
      borderRightWidth,
      borderTopEndRadius,
      borderTopLeftRadius,
      borderTopWidth,
      borderRadius,
      borderStyle,
      borderWidth,
    },
  ]
}

const generateLayoutsStyle = ({
  position,
  bottom,
  left,
  right,
  top,
  zIndex,
  overflow,
}: LayoutsProps): StyleProp<ViewStyle> => {
  return {
    position,
    left,
    bottom,
    right,
    top,
    zIndex,
    overflow,
  }
}

export const generateStyleSheet = <T extends ViewStyle | TextStyle | ImageStyle>(
  styles: StyleProp<T>[]
) => {
  const flattenStyles = StyleSheet.flatten(styles)
  const filteredFlattenStyles = removeFalsyProperties(flattenStyles)

  return filteredFlattenStyles
}

/**
 * Generates a style sheet based on the provided props and colors.
 * @param props - The styled props to generate styles for.
 * @param colors - The color palette to use for generating background and border styles.
 * @returns A style sheet object containing the generated styles.
 */
export const generateStyledSystem = (props: StyledProps, colors: Colors) => {
  const flexStyle = generateFlexStyle(props)
  const spacingStyle = generateSpacingStyle(props)
  const bgStyle = generateBgStyle(props, colors)
  const sizingStyle = generateSizingStyle(props)
  const effectsStyle = generateEffectsStyle(props)
  const bordersStyle = generateBordersStyle(props, colors)
  const layoutsStyle = generateLayoutsStyle(props)

  return generateStyleSheet([
    flexStyle,
    spacingStyle,
    bgStyle,
    sizingStyle,
    effectsStyle,
    bordersStyle,
    layoutsStyle,
  ])
}
