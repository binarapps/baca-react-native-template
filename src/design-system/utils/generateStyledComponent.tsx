import React, { forwardRef, ComponentType } from 'react'

import { generateStyledSystem } from './generateStyledSystem'
import { StyledProps } from '../components/types'

import { useTheme } from '@/hooks'

type StylePropKeys = 'style' | 'contentContainerStyle'

/**
 * A higher-order function that generates a styled component with styled-system props.
 * @template P - The props type for the component.
 * @template R - The ref type for the component.
 * @param {ComponentType<P>} component - The component to be styled.
 * @param {StylePropKeys} [styleProp='style'] - The name of the prop to pass the generated styles to (might be e.g. 'contentContainerStyle' for `ScrollView`).
 * @returns {FunctionComponent<P>} A styled component with styled-system props.
 * @example
 * import { Text } from 'react-native'
 * import { generateStyledComponent } from '@/components/utils'
 *
 * const StyledText = generateStyledComponent(Text)
 *
 *  const Component = () => (
 *    <StyledText
 *      fontSize={16}
 *      color="primary.500"
 *      fontWeight="bold"
 *    >
 *      Hello World!
 *    </StyledText>
 *  )
 */
export const generateStyledComponent = <P extends StyledProps, R>(
  component: ComponentType<P>,
  styleProp: StylePropKeys = 'style'
) =>
  forwardRef<R, P>((props, ref) => {
    const { colors } = useTheme()

    const style = generateStyledSystem(props, colors)
    return React.createElement(component, {
      ...props,
      ref,
      [styleProp]: [style, (props as Record<StylePropKeys, unknown>)[styleProp]],
    })
  })
