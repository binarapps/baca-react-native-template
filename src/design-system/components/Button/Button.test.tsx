import { Button } from './Button'

import { theme } from '@/design-system'
import { cleanup, render, fireEvent, act } from '@/utils/testUtils'

const baseStyles = {
  alignItems: 'center',
  borderRadius: 4,
  flexDirection: 'row',
  justifyContent: 'center',
  paddingHorizontal: 14,
  paddingVertical: 10,
}

afterEach(cleanup)
describe('Button', () => {
  it('renders correctly', () => {
    const { toJSON } = render(<Button title="Button" />)
    expect(toJSON()).toMatchSnapshot()
  })

  it('renders correctly Button.Primary', () => {
    const { getByTestId } = render(<Button.Primary title="Button" />)
    expect(getByTestId('baseButton').props.style).toStrictEqual({
      ...baseStyles,
      backgroundColor: theme.light.colors.button.primary.bg,
      borderColor: theme.light.colors.button.primary.border,
      borderWidth: 1,
    })
  })

  it('renders correctly Button.PrimaryDestructive', () => {
    const { getByTestId } = render(<Button.PrimaryDestructive title="Button" />)
    expect(getByTestId('baseButton').props.style).toStrictEqual({
      ...baseStyles,
      backgroundColor: theme.light.colors.button.primary.error.bg,
      borderColor: theme.light.colors.button.primary.error.border,
      borderWidth: 1,
    })
  })

  it('renders correctly Button.SecondaryColor', () => {
    const { getByTestId } = render(<Button.SecondaryColor title="Button" />)
    expect(getByTestId('baseButton').props.style).toStrictEqual({
      ...baseStyles,
      backgroundColor: theme.light.colors.button.secondary.color.bg,
      borderColor: theme.light.colors.button.secondary.color.border,
      borderWidth: 1,
    })
  })

  it('renders correctly Button.SecondaryGray', () => {
    const { getByTestId } = render(<Button.SecondaryGray title="Button" />)
    expect(getByTestId('baseButton').props.style).toStrictEqual({
      ...baseStyles,
      alignItems: 'center',
      backgroundColor: theme.light.colors.button.secondary.bg,
      borderColor: theme.light.colors.button.secondary.border,
      borderWidth: 1,
    })
  })

  it('renders correctly Button.SecondaryDestructive', () => {
    const { getByTestId } = render(<Button.SecondaryDestructive title="Button" />)
    expect(getByTestId('baseButton').props.style).toStrictEqual({
      ...baseStyles,
      backgroundColor: theme.light.colors.button.secondary.error.bg,
      borderColor: theme.light.colors.button.secondary.error.border,
      borderWidth: 1,
    })
  })

  it('renders correctly Button.TertiaryColor', () => {
    const { getByTestId } = render(<Button.TertiaryColor title="Button" />)
    expect(getByTestId('baseButton').props.style).toStrictEqual({
      ...baseStyles,
      backgroundColor: theme.light.colors.Base.transparent,
      borderColor: theme.light.colors.Base.transparent,
      borderWidth: undefined,
    })
  })

  it('renders correctly Button.TertiaryGray', () => {
    const { getByTestId } = render(<Button.TertiaryGray title="Button" />)
    expect(getByTestId('baseButton').props.style).toStrictEqual({
      ...baseStyles,
      backgroundColor: theme.light.colors.Base.transparent,
      borderColor: theme.light.colors.Base.transparent,
      borderWidth: undefined,
    })
  })

  it('renders correctly Button.TertiaryDestructive', () => {
    const { getByTestId } = render(<Button.TertiaryDestructive title="Button" />)
    expect(getByTestId('baseButton').props.style).toStrictEqual({
      ...baseStyles,
      backgroundColor: theme.light.colors.Base.transparent,
      borderColor: theme.light.colors.Base.transparent,
      borderWidth: undefined,
    })
  })

  it('renders correctly Button.LinkColor', () => {
    const { getByTestId } = render(<Button.LinkColor title="Button" />)
    expect(getByTestId('baseButton').props.style).toStrictEqual({
      ...baseStyles,
      backgroundColor: theme.light.colors.Base.transparent,
      borderColor: theme.light.colors.Base.transparent,
      borderWidth: undefined,
    })
  })

  it('renders correctly Button.LinkGray', () => {
    const { getByTestId } = render(<Button.LinkGray title="Button" />)
    expect(getByTestId('baseButton').props.style).toStrictEqual({
      ...baseStyles,
      backgroundColor: theme.light.colors.Base.transparent,
      borderColor: theme.light.colors.Base.transparent,
      borderWidth: undefined,
    })
  })

  it('renders correctly Button.LinkDestructive', () => {
    const { getByTestId } = render(<Button.LinkDestructive title="Button" />)
    expect(getByTestId('baseButton').props.style).toStrictEqual({
      ...baseStyles,
      backgroundColor: theme.light.colors.Base.transparent,
      borderColor: theme.light.colors.Base.transparent,
      borderWidth: undefined,
    })
  })

  it('should call onPress', async () => {
    const onPress = jest.fn()
    const { getByTestId, update } = render(<Button title="Button" onPress={onPress} />)
    const button = getByTestId('baseButton')
    fireEvent.press(button)
    await act(async () => {
      update(<Button />)
    })
    expect(onPress).toHaveBeenCalledTimes(1)
  })

  it('should not call onPress when disabled', async () => {
    const onPress = jest.fn()
    const { getByTestId, update } = render(<Button title="Button" disabled onPress={onPress} />)
    const button = getByTestId('baseButton')
    fireEvent.press(button)
    await act(async () => {
      update(<Button />)
    })
    expect(onPress).toHaveBeenCalledTimes(0)
  })
})
