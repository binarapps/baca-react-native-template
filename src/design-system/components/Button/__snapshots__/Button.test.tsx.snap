// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders correctly 1`] = `
<View
  accessibilityRole="button"
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  role="button"
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#7F56D9",
      "borderColor": "#7F56D9",
      "borderRadius": 4,
      "borderWidth": 1,
      "flexDirection": "row",
      "justifyContent": "center",
      "paddingHorizontal": 14,
      "paddingVertical": 10,
    }
  }
  testID="baseButton"
>
  <View
    flexDirection="row"
    gap={4}
    style={
      [
        {
          "flexDirection": "row",
          "gap": 16,
        },
        undefined,
      ]
    }
  >
    <Text
      allowFontScaling={false}
      style={
        {
          "color": "#FFFFFF",
          "fontFamily": "Inter_SemiBold",
          "fontSize": 14,
          "fontStyle": "normal",
          "fontWeight": "400",
          "lineHeight": 21,
          "textAlign": "center",
          "textTransform": "none",
        }
      }
      testID="baseText"
      textAlign="center"
    >
      Button
    </Text>
  </View>
</View>
`;
